### PR TITLE
Add kn route list command

### LINKED ISSUE #201: Implement kn route list command
Just like `kn service list` and `kn revision list`

### PR DESCRIPTION
**Pull request by navidshaikh**
Fixes #201 

Example:
```
# list all routes
kn route list
NAME    URL                                     AGE   CONDITIONS   TRAFFIC
hello   http://hello.default.apps-crc.testing   13m   1 OK / 3     100% -> hello-7w8z7
kn1     http://kn1.default.apps-crc.testing     22h   3 OK / 3     100% -> kn1-pf6jr
rt1     http://rt1.default.apps-crc.testing     22h   3 OK / 3     20% -> svc1-j7hqr, 80% -> svc1-j7hqr
```
Help section
```
kn route list --help           
List available routes.

Usage:
  kn route list NAME [flags]

Examples:

  # List all routes
  kn route list

  # List route 'web' in namespace 'dev'
  kn route list web -n dev

  # List all routes in yaml format
  kn route list -o yaml

Flags:
      --all-namespaces                If present, list the requested object(s) across all namespaces. Namespace in current context is ignored even if specified with --namespace.
      --allow-missing-template-keys   If true, ignore any errors in templates when a field or map key is missing in the template. Only applies to golang and jsonpath output formats. (default true)
  -h, --help                          help for list
  -n, --namespace string              List the requested object(s) in given namespace.
  -o, --output string                 Output format. One of: json|yaml|name|go-template|go-template-file|template|templatefile|jsonpath|jsonpath-file.
      --template string               Template string or path to template file to use when -o=go-template, -o=go-template-file. The template format is golang templates [http://golang.org/pkg/text/template/#pkg-overview].

Global Flags:
      --config string       config file (default is $HOME/.kn/config.yaml)
      --kubeconfig string   kubectl config file (default is $HOME/.kube/config)
```

### CODE REVIEW

File: pkg/kn/commands/route/list_test.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+func createMockRouteMeta(name string) *v1alpha1.Route {
+	route := &v1alpha1.Route{}
+	route.Kind = "Route"
+	route.APIVersion = "knative.dev/v1alpha1"
+	route.Name = name
+	route.Namespace = commands.FakeNamespace
+	return route
+
```
Comment (savitaashture) [inline pkg/kn/commands/route/list_test.go]: Extra line

**Comment by navidshaikh** [comment]
Failure prompts 
> 2019/06/26 07:37:17 Code coverage is below threshold (50%), fail presubmit workflow intentionally

will add more test coverage for pkg/kn/commands/route/list_flags.go

**Comment by navidshaikh** [comment]
e2e tests are failing with
```
--- FAIL: TestBasicWorkflow (148.82s)
basic_workflow_test.go:177: Expected output incorrect, expecting to include:
100% -> hello
Instead found:
NAME URL AGE CONDITIONS TRAFFIC
hello http://hello.kne2etests.example.com 21s 0 OK / 3
```

The traffic percent and revision name altogether is not showing up (I added a time.Sleep as well before executing route list command, age is 21s), also conditions for route are `0/3`, this works well on my local cluster.

**Comment by rhuss** [comment]
> The traffic percent and revision name altogether is not showing up (I added a time.Sleep as well before executing route list command, age is 21s), also conditions for route are `0/3`, this works well on my local cluster.

Let us work on #54 now and get it integrated, then hopefully we can get rid a bit of the flakiness.

File: pkg/kn/commands/route/list.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+				return err
+			}
+			if len(route.Items) == 0 {
+				fmt.Fprintf(cmd.OutOrStdout(), "No resources found.\n")
+				return nil
+			}
+			route.GetObjectKind().SetGroupVersionKind(schema.GroupVersionKind{
+				Group:   "knative.dev",
+				Version: "v1alpha1",
+				Kind:    "route"})
```
Comment (cppforlife) [inline pkg/kn/commands/route/list.go]: formatting nit: `})` should be on the new line.

Comment (cppforlife) [reply pkg/kn/commands/route/list.go]: though i see we have it like that in other list commands. we should probably fix that separately.

File: pkg/kn/commands/route/human_readable_flags.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+		name,
+		url,
+		age,
+		conditions,
+		traffic)
+	return []metav1beta1.TableRow{row}, nil
+}
+
+func calculateTraffic(targets []servingv1alpha1.TrafficTarget) string {
+	var traffic string
```
Comment (cppforlife) [inline pkg/kn/commands/route/human_readable_flags.go]: it seems that we keep on overriding traffic string in a loop?

File: pkg/kn/commands/route/list_flags_test.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+package route
+
+import (
+	"testing"
+
+	"github.com/knative/client/pkg/kn/commands"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+)
+
+func TestRoutListFlags(t *testing.T) {
```
Comment (cppforlife) [inline pkg/kn/commands/route/list_flags_test.go]: i dont think we need this test. it's too close to implementation so it doesnt add confidence that this object is correct. additionally route list flags object is used in the command test, so it gets tested there indirectly in a more resilient manner.

File: pkg/kn/commands/route/route.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+	"github.com/knative/client/pkg/kn/commands"
+	"github.com/spf13/cobra"
+)
+
+func NewRouteCommand(p *commands.KnParams) *cobra.Command {
+	routeCmd := &cobra.Command{
+		Use:   "route",
+		Short: "Route command group",
+		RunE: func(cmd *cobra.Command, args []string) error {
```
Comment (cppforlife) [inline pkg/kn/commands/route/route.go]: @maximilien i was hoping to avoid copy-pasta (and associated tests) every time we add a command.

File: pkg/kn/commands/test_helper.go [hunk 6]
```diff
@@ -67,3 +69,26 @@ Eventing: Manage event subscriptions and channels. Connect up event sources.`,
...
+
+func TestContains(t *testing.T, output string, sub []string, element string) {
+	for _, each := range sub {
+		if !strings.Contains(output, each) {
+			t.Errorf("Missing %s: %s", element, each)
+		}
+	}
+}
+
+func AreSlicesEqual(a, b []string) bool {
```
Comment (cppforlife) [inline pkg/kn/commands/test_helper.go]: this seems like could just be replaced with `reflect.DeepEqual`

File: test/e2e-smoke-tests.sh [hunk 7]
```diff
@@ -47,18 +47,16 @@ header "Running smoke tests"
 kubectl create ns $KN_E2E_SMOKE_TESTS_NAMESPACE || fail_test
 
 ./kn service create hello --image gcr.io/knative-samples/helloworld-go -e TARGET=Knative -n $KN_E2E_SMOKE_TESTS_NAMESPACE || fail_test
-sleep 5
```
Comment (cppforlife) [inline test/e2e-smoke-tests.sh]: this should probably be separated into a different PR.

File: pkg/kn/commands/route/route.go (hunk 5 above)
Comment (maximilien) [reply pkg/kn/commands/route/route.go]: Yes, dude. On this today. @navidshaikh you can remove this as #111 will fix for all command group that do not include a `RunE` and have subcommands.

File: pkg/kn/commands/route/human_readable_flags.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+	hprinters "github.com/knative/client/pkg/printers"
+	servingv1alpha1 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	metav1beta1 "k8s.io/apimachinery/pkg/apis/meta/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// RouteListHandlers adds print handlers for route list command
+func RouteListHandlers(h hprinters.PrintHandler) {
+	kRouteColumnDefinitions := []metav1beta1.TableColumnDefinition{
+		{Name: "Name", Type: "string", Description: "Name of the knative route."},
```
Comment (maximilien) [inline pkg/kn/commands/route/human_readable_flags.go]: [Nit] should we consistently refer to Knative with this spelling (capital K)? If yes, update description. Thx

File: pkg/kn/commands/route/list_flags_test.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+	routeListFlags.AddFlags(cmd)
+	printer, err := routeListFlags.ToPrinter()
+	// check if printer is compatible
+	if genericclioptions.IsNoCompatiblePrinterError(err) {
+		t.Fatalf("Expected to match human readable printer.")
+	}
+	if err != nil {
+		t.Fatalf("Failed to find a proper printer.")
+	}
+	// check if its able to printer can print the object
```
Comment (maximilien) [inline pkg/kn/commands/route/list_flags_test.go]: [Nit] part of why I hate ( 😀 ) comment for somewhat obvious code. They get out of date and can be more confusing than code or English hard to parse (as above): "check if its able to printer can print the object" what? :)

File: pkg/kn/commands/route/human_readable_flags.go (hunk 3 above)
Comment (navidshaikh) [reply pkg/kn/commands/route/human_readable_flags.go]: @cppforlife: yup but with an update only if there is traffic split available
```
traffic = fmt.Sprintf("%s, %d%% -> %s", traffic, target.Percent, target.RevisionName)
```

File: pkg/kn/commands/route/list_flags_test.go (hunk 4 above)
Comment (navidshaikh) [reply pkg/kn/commands/route/list_flags_test.go]: I added this file's test as without this `pk/kn/commands/route/list_flags.go` was failing CI go test coverage threshold (must be > 50%).

should I be testing this in some other way?

File: pkg/kn/commands/route/route.go (hunk 5 above)
Comment (navidshaikh) [reply pkg/kn/commands/route/route.go]: removed, thanks @maximilien!

File: pkg/kn/commands/test_helper.go (hunk 6 above)
Comment (navidshaikh) [reply pkg/kn/commands/test_helper.go]: thanks, updated

File: pkg/kn/commands/route/list_flags_test.go (hunk 9 above)
Comment (navidshaikh) [reply pkg/kn/commands/route/list_flags_test.go]: typo!
removed obvious code comments!!

File: test/e2e-smoke-tests.sh (hunk 7 above)
Comment (navidshaikh) [reply test/e2e-smoke-tests.sh]: sure, raised https://github.com/knative/client/pull/220

**Comment by navidshaikh** [comment]
> Can you squash the 14 commits into one? Unless you have good reason to have so many. Otherwise, looks good to me :)

The bot should do this while merging to master, for eg: check this [commit](https://github.com/knative/client/commit/65f1c081eeba5980c80d344a4886d6e5ad8396a2) squashed into one from 9.

**Comment by maximilien** [comment]
> > Can you squash the 14 commits into one? Unless you have good reason to have so many. Otherwise, looks good to me :)
> 
> The bot should do this while merging to master, for eg: check this [commit](https://github.com/knative/client/commit/65f1c081eeba5980c80d344a4886d6e5ad8396a2) squashed into one from 9.

Thanks, great to know. Best :beers

**Comment by rhuss** [comment]
> The bot should do this while merging to master, for eg: check this [commit](https://github.com/knative/client/commit/65f1c081eeba5980c80d344a4886d6e5ad8396a2) squashed into one from 9.

Tru, but you will get the sum of all commit messages (even when you did some extra rounds which then has been reverted). In that case you can still squash on your own to have a clean single commit message, too.

**Comment by rhuss** [comment]
/lgtm 

(re-adding @maximilien approval after squashing)

File: pkg/kn/commands/route/human_readable_flags.go (hunk 3 above)
Comment (cppforlife) [reply pkg/kn/commands/route/human_readable_flags.go]: im still confused, should it be `traffic += ...` instead since there could be multiple targets?
